### PR TITLE
Onion TLV support (EXPERIMENTAL_FEATURE)

### PR DESCRIPTION
**Pull request by rustyrussell**

### CODE REVIEW

File: tests/test_pay.py [hunk 1]
```diff
@@ -2331,3 +2331,70 @@ def test_error_returns_blockheight(node_factory, bitcoind):
...
+     l3.daemon.wait_for_log("Got onion.*'type': 'legacy'")
+
+     # Now we finally announce l2->l3 channel, so l3 can announce tlv support.
+     bitcoind.generate_block(2)
+     
+     wait_for(lambda: len(l1.rpc.listnodes(l3.info['id'])['nodes']) > 0)
+     wait_for(lambda: 'alias' in l1.rpc.listnodes(l3.info['id'])['nodes'][0])
+
+     inv = l3.rpc.invoice(10000, "test_tlv3", "test_tlv3")['bolt11']
+     print(inv)
```
Comment (niftynei) [inline tests/test_pay.py]: probably can delete this

File: common/sphinx.c [hunk 2]
```diff
@@ -771,3 +798,74 @@ struct onionreply *unwrap_onionreply(const tal_t *ctx,
...
+	case SPHINX_V0_PAYLOAD:
+		*amt_forward = rs->payload.v0.amt_forward;
+		*outgoing_cltv = rs->payload.v0.outgoing_cltv;
+		return true;
+	case SPHINX_TLV_PAYLOAD:
+		if (!rs->payload.tlv->amt_to_forward)
+			return false;
+		amt_forward->millisatoshis /* Raw: tu64 -> millisatoshis */
+			= rs->payload.tlv->amt_to_forward->amt_to_forward;
+		if (!rs->payload.tlv->outgoing_cltv_value)
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply common/sphinx.c]: Yes, that's better style.  Will fix.

**Comment by rustyrussell** [comment]
Rebased on top of #3245

**Comment by rustyrussell** [comment]
Back on top of master again...

File: devtools/onion.c [hunk 3]
```diff
@@ -58,49 +57,31 @@ static void do_generate(int argc, char **argv,
 			     argv[1 + i]);
 		}
 
-		memset(&hops_data[i], 0, sizeof(hops_data[i]));
-		if (argv[1 + i][klen] != '\0') {
-			/* FIXME: Generic realm support, not this hack! */
-			/* FIXME: Multi hop! */
+		/* /<hex> -> raw hopdata. /tlv -> TLV encoding. */
+		if (argv[1 + i][klen] != '\0' && argv[1 + i][klen] != 't') {
```
Comment (cdecker) [inline devtools/onion.c]: This seems to undo some of the fixes from #3250 (https://github.com/ElementsProject/lightning/pull/3250/files#diff-6ba0afe5becbb580348286dbd8ca2b5cR63). Which version is the correct one?

Comment (rustyrussell) [reply devtools/onion.c]: I've just rebased again on master, which fixes this...

File: common/sphinx.h [hunk 4]
```diff
@@ -81,6 +81,7 @@ struct route_step {
 	enum sphinx_payload_type type;
 	union {
 		struct hop_data_legacy v0;
+		struct tlv_tlv_payload *tlv;
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply common/sphinx.h]: Our marshaling/unmarshaling really wants a pointer...

File: common/sphinx.c [hunk 5]
```diff
@@ -771,3 +798,74 @@ struct onionreply *unwrap_onionreply(const tal_t *ctx,
...
+		*outgoing_cltv = rs->payload.tlv->outgoing_cltv_value->outgoing_cltv_value;
+		return true;
+	case SPHINX_INVALID_PAYLOAD:
+		return false;
+
+	/* This should probably be removed, as it's just for testing */
+	case SPHINX_RAW_PAYLOAD:
+		abort();
+	}
+	abort();
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply common/sphinx.c]: The onion tool actually uses this to make an arbitrary raw payload, so we'll leave it for a bit longer.

File: common/sphinx.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (cdecker) [reply common/sphinx.h]: Hm, ok, I should read the TLV implementation more carefully :-)

**Comment by cdecker** [comment]
ACK https://github.com/ElementsProject/lightning/pull/3167/commits/ed0b57515d46a770aef370e8d68100b1221dbb7a

**Pull request merged**